File: tests/reviewed-feedback-input.test.mjs
import test from 'node:test';
import assert from 'node:assert/strict';
import { mkdtemp, readFile, rm, writeFile } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { hydrateReviewedFeedback } from '../scripts/reviewed-feedback-input.mjs';

const fixture = JSON.parse(await readFile(new URL('../edge_repos/storyyard/tests/fixtures/inkos-source-first-pitch-review-v3.json', import.meta.url), 'utf8'));
const target = { decisionId: 'test-decision', packetId: fixture.packetId, packetSha256: fixture.packetSha256,
  candidateId: fixture.candidates[0].id, candidateSha256: fixture.candidates[0].sha256, artifactId: fixture.artifact.id };

async function withPacket(fn, packet = fixture) {
  const directory = await mkdtemp(join(tmpdir(), 'firefly-feedback-'));
  try { await writeFile(join(directory, `${fixture.packetId}.json`), JSON.stringify(packet)); return await fn({ packetDirectories: [directory] }); }
  finally { await rm(directory, { recursive: true, force: true }); }
}

test('loads the exact reviewed project plan and deduplicates multiple comments', async () => {
  await withPacket(async (options) => {
    const result = await hydrateReviewedFeedback([target, { ...target, decisionId: 'second' }], options);
    assert.equal(result.contexts.length, 1);
    assert.equal(result.contexts[0].body, fixture.candidates[0].projectPlan.markdown);
    assert.equal(result.contexts[0].representation, 'project-plan-markdown');
    assert.equal(result.contexts[0].feedbackScope, 'reviewed-candidate');
    assert.deepEqual(result.targets.map((item) => item.status), ['reviewed-text-loaded', 'reviewed-text-loaded']);
    assert.equal(result.humanReviewRequired, false);
  });
});

test('a damaged local cache cannot shadow a separately verified immutable packet copy', async () => {
  const damaged = await mkdtemp(join(tmpdir(), 'firefly-feedback-damaged-'));
  try {
    await writeFile(join(damaged, `${fixture.packetId}.json`), '{partial');
    await withPacket(async (options) => {
      const result = await hydrateReviewedFeedback([target], { packetDirectories: [damaged, ...options.packetDirectories] });
      assert.equal(result.targets[0].status, 'reviewed-text-loaded');
      assert.equal(result.contexts[0].body, fixture.candidates[0].projectPlan.markdown);
      assert.equal(await readFile(join(damaged, `${fixture.packetId}.json`), 'utf8'), '{partial');
    });
  } finally { await rm(damaged, { recursive: true, force: true }); }
});

test('candidate, packet, and artifact bindings must each match', async () => {
  await withPacket(async (options) => {
    const result = await hydrateReviewedFeedback([
      { ...target, candidateSha256: '0'.repeat(64) },
      { ...target, decisionId: 'bad-packet', packetSha256: '0'.repeat(64) },
      { ...target, decisionId: 'bad-artifact', artifactId: 'different' },
    ], options);
    assert.equal(result.contexts.length, 0);
    assert.deepEqual(result.targets.map((item) => item.status), ['candidate-binding-mismatch', 'packet-binding-mismatch', 'packet-binding-mismatch']);
  });
});

test('changed packet content is rejected even when its filename and claimed hashes match', async () => {
  const changed = structuredClone(fixture);
  changed.candidates[0].projectPlan.markdown += '\n추가된 내용';
  await withPacket(async (options) => {
    const result = await hydrateReviewedFeedback([target], options);
    assert.equal(result.contexts.length, 0);
    assert.equal(result.targets[0].status, 'packet-verification-failed');
  }, changed);
});

test('budget exclusion keeps a whole representation out instead of silently shortening it', async () => {
  await withPacket(async (options) => {
    const result = await hydrateReviewedFeedback([target], { ...options, maxCharacters: fixture.candidates[0].projectPlan.markdown.length - 1 });
    assert.equal(result.contexts.length, 0);
    assert.equal(result.characters, 0);
    assert.equal(result.targets[0].status, 'reviewed-text-omitted-budget');
  });
});

test('explicit packet feedback includes all candidates without inventing a selected candidate', async () => {
  await withPacket(async (options) => {
    const result = await hydrateReviewedFeedback([{ ...target, candidateId: '', candidateSha256: '' }], { ...options, maxCharacters: 120000 });
    assert.equal(result.contexts[0].feedbackScope, 'whole-reviewed-packet');
    assert.equal(result.contexts[0].candidateId, '');
    const body = JSON.parse(result.contexts[0].body);
    assert.deepEqual(body.candidates.map((item) => item.id), fixture.candidates.map((item) => item.id));
  });
});

test('missing packets and malformed bindings remain explicit and empty history does no lookup', async () => {
  await withPacket(async (options) => {
    const result = await hydrateReviewedFeedback([
      { ...target, packetId: `frp-${'0'.repeat(24)}` },
      { ...target, decisionId: 'malformed', candidateId: '../secret' },
    ], options);
    assert.deepEqual(result.targets.map((item) => item.status), ['packet-not-found', 'missing-or-invalid-binding']);
  });
  assert.equal((await hydrateReviewedFeedback([], { root: '/nonexistent' })).contexts.length, 0);
});

File: tests/reviewed-feedback-input.test.py
import importlib.util
import json
from pathlib import Path
import unittest
from unittest.mock import patch

ROOT = Path(__file__).resolve().parents[1]
spec = importlib.util.spec_from_file_location("feedback_test", ROOT / "scripts/reviewed-feedback-input.py")
feedback = importlib.util.module_from_spec(spec)
spec.loader.exec_module(feedback)


def fixture():
    row = {"id": "d1", "packetId": "frp-" + "a" * 24, "packetSha256": "a" * 64,
           "candidateId": "p01", "candidateSha256": "b" * 64, "artifactId": "pitch",
           "decision": "hold", "comment": "이 장면의 거래 목적이 불명확하다.",
           "createdAt": "2026-09-12T00:00:00Z", "status": "pending", "actorEmail": "private@example.test"}
    scope = feedback.planning.scope_hil([row])
    context = {"id": "context-1", "packetId": row["packetId"], "packetSha256": row["packetSha256"],
               "candidateId": "p01", "candidateSha256": row["candidateSha256"], "artifactId": "pitch",
               "body": "원래 검토한 기획 본문", "contentSha256": feedback.planning.digest("원래 검토한 기획 본문")}
    hydrated = {"contexts": [context], "characters": len(context["body"]), "targets": [
        {"decisionId": "d1", "packetId": row["packetId"], "candidateId": "p01", "status": "reviewed-text-loaded", "contextId": "context-1"}]}
    return scope, hydrated


class FeedbackTests(unittest.TestCase):
    def test_attaches_content_to_exact_candidate_and_keeps_raw_history_unchanged(self):
        scope, hydrated = fixture()
        before = json.dumps(scope, ensure_ascii=False)
        result = feedback.planning.attach_reviewed_feedback(scope, hydrated)
        self.assertEqual(json.dumps(scope, ensure_ascii=False), before)
        self.assertEqual(result["receipt"]["rawDecisions"], scope["receipt"]["rawDecisions"])
        self.assertEqual(result["writerContext"]["candidateFeedback"][0]["targets"][0]["contextId"], "context-1")
        self.assertIn("새 사람 검토를 기다릴 필요가 없다", result["writerContext"]["application"])
        self.assertNotIn("private@example.test", json.dumps(result["writerContext"]))
        self.assertEqual(result["receipt"]["writerContextSha256"], feedback.planning.digest(json.dumps(result["writerContext"], ensure_ascii=False)))

    def test_invalid_content_or_binding_cannot_be_labelled_reviewed(self):
        for key, value in (("body", "바뀐 본문"), ("candidateSha256", "c" * 64), ("artifactId", "other")):
            with self.subTest(key=key):
                scope, hydrated = fixture()
                hydrated["contexts"][0][key] = value
                with self.assertRaises(ValueError):
                    feedback.planning.attach_reviewed_feedback(scope, hydrated)
        scope, hydrated = fixture()
        hydrated["targets"] = []
        with self.assertRaises(ValueError):
            feedback.planning.attach_reviewed_feedback(scope, hydrated)

    def test_missing_local_runtime_records_an_explicit_gap_without_blocking_work(self):
        scope, _ = fixture()
        with patch.object(feedback.subprocess, "run", side_effect=FileNotFoundError("node unavailable")):
            result = feedback.load_reviewed_feedback(scope)
        self.assertEqual(result["writerContext"]["reviewedContexts"], [])
        self.assertEqual(result["writerContext"]["candidateFeedback"][0]["targets"][0]["contextStatus"], "context-loader-unavailable")
        self.assertFalse(result["receipt"]["reviewedContext"]["humanReviewRequired"])
        with patch.object(feedback.subprocess, "run", side_effect=AssertionError("No lookup needed")):
            empty = feedback.planning.scope_hil([])
            self.assertIs(feedback.load_reviewed_feedback(empty), empty)


if __name__ == "__main__":
    unittest.main()
